### ISSUE TITLE
Optimising with respect to the epoch that scored highest for a trial (instead of the last epoch)

### ISSUE THREAD

**Issue by anicolson** [ray-project/ray_lightning#81]
Hi, 

I am using `metric="acc"` for `ray.tune`, with `mode="max"`. However, I think that the score for the last epoch is being used as the "best" score for the trial.

E.g., for trial `train_mnist_c1a02_00000 ` from the following console output, the reported `acc` is `0.942736`:

```
+-------------------------+------------+-------+--------------+-----------+-----------+-------------+--------+------------------+----------+----------+
| Trial name              | status     | loc   |   batch_size |   layer_1 |   layer_2 |          lr |   iter |   total time (s) |      loss |      acc |
|-------------------------+------------+-------+--------------+-----------+-----------+-------------+--------+------------------+-----------+----------|
| train_mnist_c1a02_00000 | TERMINATED |       |           64 |       128 |       256 | 0.000120742 |     16 |    166.36  | -0.938587 | 0.942736 |
| train_mnist_c1a02_00001 | TERMINATED |       |          128 |       128 |        64 | 0.000120068 |     16 |    138.23  | -0.923084 | 0.929161 |
| train_mnist_c1a02_00002 | TERMINATED |       |           64 |        32 |       256 | 0.000308457 |     16 |    168.73  | -0.942267 | 0.945811 |
| train_mnist_c1a02_00003 | TERMINATED |       |           64 |        32 |       256 | 0.0927983   |     16 |    162.749 | -0.103807 | 0.103807 |
+-------------------------+------------+-------+--------------+-----------+-----------+-------------+--------+------------------+-----------+----------+
```

However, when looking at `progress.csv` in `tune_mnist/train_mnist_c1a02_00000_0_batch_size=64,layer_1=128,layer_2=256,lr=0.00012074_2021-08-25_20-18-14`, the highest `acc` is `0.9441488981246948` (where `0.942736` is the score for the last epoch):

```
loss,acc,time_this_iter_s,done,timesteps_total,episodes_total,training_iteration,experiment_id,date,timestamp,time_total_s,pid,hostname,node_ip,time_since_restore,timesteps_since_restore,iterations_since_restore,trial_id
-0.7927070260047913,0.8189826607704163,28.36010980606079,False,,,1,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-18-49,1629886729,28.36010980606079,6438,b078,10.141.1.144,28.36010980606079,0,1,c1a02_00000
-0.8159223198890686,0.829454779624939,9.054965734481812,False,,,2,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-18-58,1629886738,37.4150755405426,6438,b078,10.141.1.144,37.4150755405426,0,2,c1a02_00000
-0.8259921669960022,0.835106372833252,9.04287576675415,False,,,3,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-19-07,1629886747,46.45795130729675,6438,b078,10.141.1.144,46.45795130729675,0,3,c1a02_00000
-0.831539511680603,0.8390957117080688,8.813407182693481,False,,,4,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-19-16,1629886756,55.271358489990234,6438,b078,10.141.1.144,55.271358489990234,0,4,c1a02_00000
-0.8353389501571655,0.8413397073745728,9.657632112503052,False,,,5,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-19-26,1629886766,64.92899060249329,6438,b078,10.141.1.144,64.92899060249329,0,5,c1a02_00000
-0.896843671798706,0.9082446694374084,9.233526706695557,False,,,6,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-19-35,1629886775,74.16251730918884,6438,b078,10.141.1.144,74.16251730918884,0,6,c1a02_00000
-0.9138997197151184,0.9224567413330078,9.06982707977295,False,,,7,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-19-44,1629886784,83.23234438896179,6438,b078,10.141.1.144,83.23234438896179,0,7,c1a02_00000
-0.919984757900238,0.9273603558540344,9.12305760383606,False,,,8,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-19-53,1629886793,92.35540199279785,6438,b078,10.141.1.144,92.35540199279785,0,8,c1a02_00000
-0.9245654940605164,0.9311003684997559,9.59031629562378,False,,,9,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-03,1629886803,101.94571828842163,6438,b078,10.141.1.144,101.94571828842163,0,9,c1a02_00000
-0.9273290038108826,0.9330950379371643,9.259892702102661,False,,,10,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-12,1629886812,111.20561099052429,6438,b078,10.141.1.144,111.20561099052429,0,10,c1a02_00000
-0.9301624298095703,0.935339093208313,9.178678750991821,False,,,11,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-21,1629886821,120.38428974151611,6438,b078,10.141.1.144,120.38428974151611,0,11,c1a02_00000
-0.9327710270881653,0.9389959573745728,9.190800189971924,False,,,12,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-30,1629886830,129.57508993148804,6438,b078,10.141.1.144,129.57508993148804,0,12,c1a02_00000
-0.9332801103591919,0.9382479786872864,9.453409433364868,False,,,13,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-40,1629886840,139.0284993648529,6438,b078,10.141.1.144,139.0284993648529,0,13,c1a02_00000
-0.9350691437721252,0.9409075379371643,8.920868873596191,False,,,14,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-49,1629886849,147.9493682384491,6438,b078,10.141.1.144,147.9493682384491,0,14,c1a02_00000
-0.9383015632629395,0.9441488981246948,9.268950700759888,False,,,15,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-20-58,1629886858,157.21831893920898,6438,b078,10.141.1.144,157.21831893920898,0,15,c1a02_00000
-0.9385870099067688,0.942736029624939,9.141263961791992,False,,,16,daa92b03869b42409a0f9a69b6d7918d,2021-08-25_20-21-07,1629886867,166.35958290100098,6438,b078,10.141.1.144,166.35958290100098,0,16,c1a02_00000

```

I see that this was an issue here: https://github.com/ray-project/ray/issues/5174, but it isn't clear what the fix was.

Thanks.

Source:
```
"""Simple example using RayAccelerator and Ray Tune"""

from pl_bolts.datamodules.mnist_datamodule import MNISTDataModule
from ray import tune
from ray_lightning.tests.utils import LightningMNISTClassifier
from ray_lightning.tune import TuneReportCallback, get_tune_ddp_resources
from ray_lightning import RayPlugin
import os
import pytorch_lightning as pl
import ray

DATA_DIR = "/datasets/work/hb-mlaifsp-mm/source/Datasets/mnist"
NUM_WORKERS = 1
NUM_SAMPLES = 4
MAX_EPOCHS = 16
USE_GPU = True

def train_mnist(config):


    model = LightningMNISTClassifier(config, DATA_DIR)

    metrics = {"loss": "ptl/val_loss", "acc": "ptl/val_accuracy"}
    callbacks = [TuneReportCallback(metrics, on="validation_end")]

    trainer = pl.Trainer(
        max_epochs=MAX_EPOCHS,
        callbacks=callbacks,
        progress_bar_refresh_rate=0,
        plugins=[RayPlugin(num_workers=NUM_WORKERS, use_gpu=USE_GPU)],
    )

    dm = MNISTDataModule(data_dir=DATA_DIR, num_workers=NUM_WORKERS, batch_size=config["batch_size"])
    trainer.fit(model, dm)

def tune_mnist():
    config = {
        "layer_1": tune.choice([32, 64, 128]),
        "layer_2": tune.choice([64, 128, 256]),
        "lr": tune.loguniform(1e-4, 1e-1),
        "batch_size": tune.choice([32, 64, 128]),
    }

    ray.init()
    analysis = tune.run(
        train_mnist,
        metric="acc",
        mode="max",
        local_dir=os.getcwd(),
        config=config,
        num_samples=NUM_SAMPLES,
        resources_per_trial=get_tune_ddp_resources(num_workers=NUM_WORKERS, use_gpu=USE_GPU),
        name="tune_mnist",
    )
    print("Best hyperparameters found were: ", analysis.best_config)

if __name__ == "__main__":
    tune_mnist()
```

**Comment by addisonklinke** [comment]
@anicolson Did you come up with a final solution? From the most recent comment, it sounds like things were still a bit up in the air
